File: Dart/web/Model/GUI.dart
import 'dart:html';
import 'Database.dart';
import 'Settings.dart';
import 'State.dart';
import 'TList.dart';
import 'TStep.dart';
import 'Todo.dart';

enum eSidebarState { Opened, Closed }

enum eDialogContent { Text, Settings, TListSettings }

class GUI {
  static final GUI Current = GUI._gui();

  factory GUI() {
    return Current;
  }

  GUI._gui();

  bool SearchView = false;
  eSidebarState LeftSidebarStatus = eSidebarState.Closed;
  eSidebarState RightSidebarStatus = eSidebarState.Closed;
  eDialogContent DialogContent = eDialogContent.Text;

  void Initialize() {
    SetColorMode();
    baseInitialize();
  }

  void baseInitialize() {
    querySelector('#btnHideLeftSidebar').onClick.listen((e) {
      ToggleLeftSidebar();
    });
    querySelector('#btnLeftSidebarSettings').onClick.listen((e) {
      ShowSettingsDialog();
    });
    querySelector('#fileLeftSidebarImport').onChange.listen((e) {
      State.Current.ImportData(e.target);
    });
    querySelector('#btnLeftSidebarExport').onClick.listen((e) {
      State.Current.ExportData();
    });
    querySelector('#btnLeftSidebarSearch').onClick.listen((e) {
      State.Current.ShowSearchView();
    });
    querySelector('#txtNewTListText').onKeyUp.listen((e) {
      State.Current.AddNewTListTextKeyUp(e);
    });
    querySelector('#btnAddNewTList').onClick.listen((e) {
      State.Current.AddNewTList();
    });

    if (LeftSidebarStatus == eSidebarState.Closed) {
      querySelector('#LeftSidebar').classes.add('is-hidden');
    }
    if (RightSidebarStatus == eSidebarState.Closed) {
      querySelector('#RightSidebar').classes.add('is-hidden');
    }

    querySelector('#txtRightSidebarTodoText').onChange.listen((e) {
      State.Current.TodoTextChanged(e.target);
    });
    querySelector('#btnAddNewTStep').onClick.listen((e) {
      State.Current.AddTStep();
    });
    querySelector('#txtNewTStepText').onKeyUp.listen((e) {
      State.Current.AddNewTStepTextKeyUp(e);
    });
    querySelector('#dtpRightSidebarTodoEndDate').onChange.listen((e) {
      State.Current.TodoEndDateChanged(e.target);
    });
    querySelector('#txaRightSidebarTodoNote').onChange.listen((e) {
      State.Current.TodoNoteChanged(e.target);
    });
    querySelector('#btnHideRightSidebar').onClick.listen((e) {
      State.Current.HideRightSidebar();
    });
    querySelector('#btnDeleteTodo').onClick.listen((e) {
      State.Current.RemoveTodo();
    });
    querySelector('#btnToggleLeftSidebar').onClick.listen((e) {
      ToggleLeftSidebar();
    });
    querySelector('#txtHeaderTListName').onChange.listen((e) {
      State.Current.TListNameChange(e.target);
    });
    querySelector('#btnDeleteTList').onClick.listen((e) {
      State.Current.RemoveTList();
    });
    querySelector('#btnSettingsTList').onClick.listen((e) {
      ShowTListSettingsDialog();
    });
    querySelector('#txtSearchText').onInput.listen((e) {
      State.Current.Search();
    });
    querySelector('#Main').onClick.listen((e) {
      OutsideClick();
    });
    querySelector('#Main').onResize.listen((e) {
      GetContentCoverStyle();
    });
    querySelector('#btnAddNewTodo').onClick.listen((e) {
      State.Current.AddTodo();
    });
    querySelector('#txtNewTodoText').onKeyUp.listen((e) {
      State.Current.AddNewTodoTextKeyUp(e);
    });
    querySelector('#btnHideDialog').onClick.listen((e) {
      HideDialog();
    });
    querySelector('#slbIcon').onChange.listen((e) {
      State.Current.TListIconChange(e.target);
    });
    querySelector('#slbOrderTodo').onChange.listen((e) {
      State.Current.TListOrderTodoChange(e.target);
    });
    querySelector('#chbDarkMode').onChange.listen((e) {
      State.Current.SetColorMode(e.target);
    });
    HideSearchView();
    querySelector('#chbRightSidebarTodoDone').onClick.listen((e) {
      State.Current.TodoDoneChanged(State.Current.CurrentTodo.Id);
    });
    querySelector('#btnRightSidebarTodoImportant').onClick.listen((e) {
      State.Current.TodoImportantChanged(State.Current.CurrentTodo.Id);
    });
  }

  void DrawTLists() {
    querySelector('#NavigationBaseTLists').children.clear();
    querySelector('#NavigationUserTLists').children.clear();
    CreateTList(
        State.Current.TodayTodos, 'lstTodayTodos', 'NavigationBaseTLists');
    CreateTList(State.Current.ImportantTodos, 'lstImportantTodos',
        'NavigationBaseTLists');
    CreateTList(
        State.Current.PlannedTodos, 'lstPlannedTodos', 'NavigationBaseTLists');
    CreateTList(State.Current.UndividedTodos, 'lstUndividedTodos',
        'NavigationBaseTLists');

    for (var item in State.Current.TLists) {
      CreateTList(item, 'lstItem${item.Id}', 'NavigationUserTLists');
    }
  }

  void CreateTList(TList tList, String lstId, String container, String result) {
    var li = LIElement()
      ..id = lstId
      ..onClick.listen((e) {
        State.Current.SwitchTList(tList);
      })
      ..classes.add('LeftSidebar-item');
    if (State.Current.CurrentTList == tList) {
      li.classes.add('is-selected');
    }
    var spanIcon = SpanElement()
      ..className = 'LeftSidebar-itemIcon ' + tList.Icon;
    li.children.add(spanIcon);
    var spanText = SpanElement()
      ..classes.add('LeftSidebar-itemText')
      ..text = tList.Name;
    li.children.add(spanText);
    var spanCount = SpanElement()
      ..classes.add('LeftSidebar-itemCount')
      ..text = result;
    li.children.add(spanCount);
    querySelector('#${container}').children.add(li);
  }

  void SetRightSidebar() {
    (querySelector('#chbRightSidebarTodoDone') as InputElement).checked =
        State.Current.CurrentTodo.Done;
    (querySelector('#txtRightSidebarTodoText') as InputElement).value =
        State.Current.CurrentTodo.Text;
    SetRightSidebarTodoImportant(State.Current.CurrentTodo.Important);
    (querySelector('#dtpRightSidebarTodoEndDate') as InputElement).value =
        State.Current.CurrentTodo.EndDate != DateMaxValue
            ? '${State.Current.CurrentTodo.EndDate}'
            : ' ';
    (querySelector('#txaRightSidebarTodoNote') as InputElement).value =
        State.Current.CurrentTodo.Note;

    querySelector('#lblRightSidebarTodoCreateDate').text =
        State.Current.CurrentTodo.CreateDate;
  }

  void DrawMain() {
    querySelector('#TodoDividerContainer').children.clear();
    if (SearchView) {
      CreateTodoDivider('dvcSearched', 'Vyhledáno', true,
          State.Current.UnfinishedTodos, true);
    } else {
      CreateTodoDivider(
          'dvcUnfinished',
          'Nedokončeno',
          State.Current.CurrentTList.UnfinishedOpened,
          State.Current.UnfinishedTodos,
          true);
      CreateTodoDivider(
          'dvcFinished',
          'Dokončeno',
          State.Current.CurrentTList.FinishedOpened,
          State.Current.FinishedTodos,
          false);
    }
  }

  void CreateTodoDivider(String id, String text, bool opened, List<Todo> todos,
      bool unfinishedButton) {
    var dividerContainerId = '${id}TodoContainer';
    var divDivider = DivElement();
    divDivider.id = id;
    divDivider.onClick.listen((event) {
      State.Current.TodoDividerClick(unfinishedButton);
    });
    divDivider.classes.add('TaskDivider');
    var spanBtn = SpanElement();
    if (opened) {
      spanBtn.classes.add('bi bi-chevron-down');
    } else {
      spanBtn.classes.add('bi bi-chevron-right');
    }
    divDivider.children.add(spanBtn);
    var spanText = SpanElement();
    spanText.text = text;
    divDivider.children.add(spanText);
    var spanCount = SpanElement();
    spanCount.text =
        (todos != null && todos.isNotEmpty) ? todos.length.toString() : ' ';
    divDivider.children.add(spanCount);
    var divContainer = DivElement();
    divContainer.id = dividerContainerId;
    querySelector('#TodoDividerContainer').children.add(divDivider);
    querySelector('#TodoDividerContainer').children.add(divContainer);

    if (todos != null && opened) {
      for (var todo in todos) {
        CreateTodo(todo, dividerContainerId);
      }
    }
  }

  void CreateTodo(Todo todo, String container) {
    var divBlock = DivElement();
    divBlock.classes.add('TaskBlock');
    if (State.Current.CurrentTodo == todo) {
      divBlock.classes.add('is-selected');
    }

    var divCheck = DivElement();
    divCheck.classes.add('TaskBlock-check');
    divBlock.children.add(divCheck);

    var divCheckContainer = DivElement();
    divCheckContainer.classes.add('CheckboxContainer');
    divCheck.children.add(divCheckContainer);

    var divRound = DivElement();
    divRound.classes.add('Round');
    divCheckContainer.children.add(divRound);

    var inputDone = InputElement();
    inputDone.id = '#chbTaskDone${todo.Id}';
    inputDone.onClick.listen((event) {
      State.Current.TodoDoneChanged(todo.Id);
    });
    if (todo.Done) {
      inputDone.classes.add('checked');
    }
    inputDone.type = 'checkbox';
    divRound.children.add(inputDone);

    var labelDone = LabelElement();
    labelDone.htmlFor = '#chbTaskDone${todo.Id}';
    divRound.children.add(labelDone);

    var spanText = SpanElement();
    spanText.id = 'lblTaskText${todo.Id}';
    spanText.classes.add('TaskBlock-text');
    spanText.onClick.listen((event) {
      State.Current.TodoSelected(todo.Id);
    });
    spanText.text = todo.Text;
    divBlock.children.add(spanText);

    var spanImportant = SpanElement();
    spanImportant.id = 'lblTaskImportant${todo.Id}';
    spanImportant.classes.add('TaskBlock-important');
    if (todo.Important) {
      spanImportant.classes.add('is-checked');
    }
    spanImportant.onClick.listen((event) {
      State.Current.TodoImportantChanged(todo.Id);
    });
    divBlock.children.add(spanImportant);

    var spanIcon = SpanElement();
    spanIcon.classes.add('TaskBlock-importantIcon');
    spanIcon.classes.add('bi-star-fill');
    spanImportant.children.add(spanIcon);

    querySelector('#${container}').children.add(divBlock);
  }

  void DrawTSteps() {
    querySelector('#TStepContainer').children.clear();
    for (var step in State.Current.TSteps) {
      CreateTStep(step);
    }
  }

  void CreateTStep(TStep step) {
    var divStep = DivElement();
    divStep.classes.add('RightSidebar-step');

    var divCheck = DivElement();
    divCheck.classes.add('RightSidebar-taskCheck');
    divStep.children.add(divCheck);

    var divCheckContainer = DivElement();
    divCheckContainer.classes.add('CheckboxContainer');
    divCheck.children.add(divCheckContainer);

    var divRound = DivElement();
    divRound.classes.add('Round');
    divCheckContainer.children.add(divRound);

    var inputDone = InputElement();
    inputDone.id = '#chbStepDone${step.Id}';
    inputDone.onClick.listen((event) {
      State.Current.TStepCheckedChanged(step.Id);
    });
    if (step.Done) {
      inputDone.classes.add('checked');
    }
    inputDone.type = 'checkbox';
    divRound.children.add(inputDone);

    var labelDone = LabelElement();
    labelDone.htmlFor = '#chbStepDone${step.Id}';
    divRound.children.add(labelDone);

    var spanTextbox = SpanElement();
    spanTextbox.classes.add('RightSidebar-taskText');
    spanTextbox.classes.add('TextBox');
    divStep.children.add(spanTextbox);

    var inputText = InputElement();
    inputText.id = 'txtStepText${step.Id}';
    inputText.classes.add('TextBox-input');
    inputText.onChange.listen((e) {
      State.Current.TStepTextChanged(e.target, step.Id);
    });
    inputText.value = step.Text;
    inputText.type = 'text';
    spanTextbox.children.add(inputText);

    var btnImportant = ButtonElement()
      ..id = 'btnStepImportant${step.Id}'
      ..onClick.listen((event) {
        State.Current.RemoveTStep(step.Id);
      })
      ..classes.add('Button')
      ..classes.add('RightSidebar-taskDelete')
      ..title = 'Smazat krok';
    divStep.children.add(btnImportant);

    var spanIcon = SpanElement();
    spanIcon.classes.add('bi bi-trash');
    btnImportant.children.add(spanIcon);

    querySelector('#TStepContainer').children.add(divStep);
  }

  void SetHeader() {
    querySelector('#lblHeaderTListIcon').classes.clear();
    querySelector('#lblHeaderTListIcon')
        .classes
        .add(State.Current.CurrentTList.Icon);
    if (SearchView) {
      ShowSearchView();
    } else {
      HideSearchView();
    }
    (querySelector('#txtHeaderTListName') as InputElement).readOnly =
        State.Current.CurrentTList.Id == UndividedTListId;
    (querySelector('#txtHeaderTListName') as InputElement).value =
        State.Current.CurrentTList.Name;
    if (State.Current.CurrentTList.Id != UndividedTListId) {
      querySelector('#TListControls').classes.remove('u-d-none');
    } else {
      querySelector('#TListControls').classes.add('u-d-none');
    }
    (querySelector('#txtSearchText') as InputElement).value = '';
  }

  void SetColorMode() {
    if (Settings.Current.GetDarkMode() == 'true') {
      if (querySelector('#Page').classes.contains('LightMode')) {
        querySelector('#Page').classes.remove('LightMode');
      }
      querySelector('#Page').classes.add('DarkMode');
    } else {
      if (querySelector('#Page').classes.contains('DarkMode')) {
        querySelector('#Page').classes.remove('DarkMode');
      }
      querySelector('#Page').classes.add('LightMode');
    }
  }

  void GetContentCoverStyle() {
    if (IsSmallScreen() &&
        (LeftSidebarStatus == eSidebarState.Opened ||
            RightSidebarStatus == eSidebarState.Opened)) {
      querySelector('#MainCover').classes.add('Main-coverContent');
    } else {
      querySelector('#MainCover').classes.remove('Main-coverContent');
    }
  }

  void SetRightSidebarTodoImportant(todo) {
    if (todo.Important) {
      querySelector('#btnRightSidebarTodoImportant').classes.add('is-checked');
    } else {
      querySelector('#btnRightSidebarTodoImportant')
          .classes
          .remove('is-checked');
    }
  }

  void OutsideClick() {
    if (IsSmallScreen()) {
      querySelector('#LeftSidebar').classes.add('is-hidden');
      LeftSidebarStatus = eSidebarState.Closed;
      GetContentCoverStyle();
    }
  }

  void ToggleLeftSidebar() {
    LeftSidebarStatus = ToggleSidebar(LeftSidebarStatus, '#LeftSidebar');
    GetContentCoverStyle();
  }

  void ToggleRightSidebar() {
    RightSidebarStatus = ToggleSidebar(RightSidebarStatus, '#RightSidebar');
    GetContentCoverStyle();
  }

  eSidebarState ToggleSidebar(eSidebarState sidebarStatus, String id) {
    switch (sidebarStatus) {
      case eSidebarState.Closed:
        querySelector(id).classes.remove('is-hidden');
        return eSidebarState.Opened;
      default:
        querySelector(id).classes.add('is-hidden');
        return eSidebarState.Closed;
    }
  }

  bool IsSmallScreen() {
    return querySelector('#Page').clientWidth < 992;
  }

  /*void DownloadFile(String fileBase64) {
        fetch('data:application/json;base64,${fileBase64}')
            .then(response => response.blob())
            .then(blob => {
                let link = window.document.createElement('a');
                link.href = window.URL.createObjectURL(blob);
                link.download = 'TodoExport.json';
                document.body.appendChild(link);
                link.click();
                document.body.removeChild(link);
            });
    }*/

  void ShowSearchView() {
    querySelector('#lblHeaderTListIcon').classes.add('u-d-none');
    querySelector('#lblHeaderTListIcon').classes.add('u-d-none');
    querySelector('#txtHeaderTListName').classes.add('u-d-none');
    querySelector('#TListControls').classes.add('u-d-none');
    querySelector('#SearchViewEditor').classes.remove('u-d-none');
    querySelector('#AddTodoContainer').classes.add('u-d-none');
    (querySelector('#txtSearchText') as InputElement).value = '';
    SearchView = true;
  }

  void HideSearchView() {
    querySelector('#lblHeaderTListIcon').classes.remove('u-d-none');
    querySelector('#lblHeaderTListIcon').classes.remove('u-d-none');
    querySelector('#txtHeaderTListName').classes.remove('u-d-none');
    querySelector('#TListControls').classes.remove('u-d-none');
    querySelector('#SearchViewEditor').classes.add('u-d-none');
    querySelector('#AddTodoContainer').classes.remove('u-d-none');
    SearchView = false;
  }

  String GetSearchText() {
    return (querySelector('#txtSearchText') as InputElement).value;
  }

  void ClearNewTListName() {
    (querySelector('#txtNewTListText') as InputElement).value = '';
  }

  String GetNewTListName() {
    return (querySelector('#txtNewTListText') as InputElement).value;
  }

  void ClearNewTodoText() {
    (querySelector('#txtNewTodoText') as InputElement).value = '';
  }

  String GetNewTodoText() {
    return (querySelector('#txtNewTodoText') as InputElement).value;
  }

  void ClearNewTStepText() {
    (querySelector('#txtNewTStepText') as InputElement).value = '';
  }

  String GetNewTStepText() {
    return (querySelector('#txtNewTStepText') as InputElement).value;
  }

  void ShowTextDialog(String title, String text) {
    DialogContent = eDialogContent.Text;
    querySelector('#lblDialogHead').text = title;
    querySelector('#DialogTextContainer').text = text;
    ShowDialog();
  }

  void ShowSettingsDialog() {
    DialogContent = eDialogContent.Settings;
    querySelector('#lblDialogHead').text = 'Nastavení';
    ShowDialog();
  }

  void ShowTListSettingsDialog() {
    DialogContent = eDialogContent.TListSettings;
    querySelector('#lblDialogHead').text =
        'Nastavení seznamu: ' + State.Current.CurrentTList.Name;
    ShowDialog();
  }

  void ShowDialog() {
    switch (DialogContent) {
      case eDialogContent.Settings:
        querySelector('#DialogTextContainer').classes.add('u-d-none');
        querySelector('#DialogSettingsContainer').classes.remove('u-d-none');
        querySelector('#DialogTListSettingsContainer').classes.add('u-d-none');
        (querySelector('#chbDarkMode') as InputElement).checked =
            Settings.Current.GetDarkMode() == 'true';
        break;
      case eDialogContent.TListSettings:
        querySelector('#DialogTextContainer').classes.add('u-d-none');
        querySelector('#DialogSettingsContainer').classes.add('u-d-none');
        querySelector('#DialogTListSettingsContainer')
            .classes
            .remove('u-d-none');
        (querySelector('#slbOrderTodo') as InputElement).value =
            State.Current.CurrentTList.OrderTodo.index.toString();
        (querySelector('#slbIcon') as InputElement).value =
            State.Current.CurrentTList.Icon;
        break;
      default:
        querySelector('#DialogTextContainer').classes.remove('u-d-none');
        querySelector('#DialogSettingsContainer').classes.add('u-d-none');
        querySelector('#DialogTListSettingsContainer').classes.add('u-d-none');
        break;
    }
    querySelector('#Dialog').classes.add('is-showed');
  }

  void HideDialog() {
    querySelector('#Dialog').classes.remove('is-showed');
  }
}

File: TypeScript/TSTodo/wwwroot/Model/GUI.ts
/// <reference path="jquery.d.ts" />

enum SidebarState {
    Opened = 0,
    Closed
}

enum DialogContent {
    Text = 0,
    Settings,
    TListSettings
}

class GUI {
    private static gui: GUI;
    private constructor() { }

    public static Current(): GUI {
        if (!this.gui) {
            this.gui = new GUI();
        }
        return this.gui;
    }

    public SearchView: boolean = false;
    public LeftSidebarStatus: SidebarState = SidebarState.Closed;
    public RightSidebarStatus: SidebarState = SidebarState.Closed;
    public DialogContent: DialogContent = DialogContent.Text;

    public Initialize(): void {
        this.SetColorMode();
        this.BaseInitialize();
    }

    public BaseInitialize(): void {
        $("#btnHideLeftSidebar").click(() => { this.ToggleLeftSidebar(); });
        $("#btnLeftSidebarSettings").click(() => { this.ShowSettingsDialog(); });
        $("#fileLeftSidebarImport").change((e) => { State.Current().ImportData(e.target); });
        $("#btnLeftSidebarExport").click(() => { State.Current().ExportData(); });
        $("#btnLeftSidebarSearch").click(() => { State.Current().ShowSearchView(); });
        $("#txtNewTListText").keyup((e) => { State.Current().AddNewTListTextKeyUp(e); });
        $("#btnAddNewTList").click(() => { State.Current().AddNewTList(); });
        if (this.LeftSidebarStatus == SidebarState.Closed)
            $("#LeftSidebar").addClass("is-hidden")
        if (this.RightSidebarStatus == SidebarState.Closed)
            $("#RightSidebar").addClass("is-hidden")
        $("#txtRightSidebarTodoText").change((e) => { State.Current().TodoTextChanged(e.target); });
        $("#btnAddNewTStep").click(() => { State.Current().AddTStep(); });
        $("#txtNewTStepText").keyup((e) => { State.Current().AddNewTStepTextKeyUp(e); });
        $("#dtpRightSidebarTodoEndDate").change((e) => { State.Current().TodoEndDateChanged(e.target); });
        $("#txaRightSidebarTodoNote").change((e) => { State.Current().TodoNoteChanged(e.target); });
        $("#btnHideRightSidebar").click(() => { State.Current().HideRightSidebar(); });
        $("#btnDeleteTodo").click(() => { State.Current().RemoveTodo(); });
        $("#btnToggleLeftSidebar").click(() => { this.ToggleLeftSidebar(); });
        $("#txtHeaderTListName").change((e) => { State.Current().TListNameChange(e.target); });
        $("#btnDeleteTList").click(() => { State.Current().RemoveTList(); });
        $("#btnSettingsTList").click(() => { this.ShowTListSettingsDialog(); });
        $("#txtSearchText").on("input", () => { State.Current().Search(); });
        $("#Main").click(() => { this.OutsideClick(); });
        $("#Main").resize(() => { this.GetContentCoverStyle(); });
        $("#btnAddNewTodo").click(() => { State.Current().AddTodo(); });
        $("#txtNewTodoText").keyup((e) => { State.Current().AddNewTodoTextKeyUp(e); });
        $("#btnHideDialog").click(() => { this.HideDialog(); });
        $("#slbIcon").change((e) => { State.Current().TListIconChange(e.target); });
        $("#slbOrderTodo").change((e) => { State.Current().TListOrderTodoChange(e.target); });
        $("#chbDarkMode").change((e) => { State.Current().SetColorMode(e.target); });
        this.HideSearchView();
        $("#chbRightSidebarTodoDone").click(() => { State.Current().TodoDoneChanged(State.Current().CurrentTodo.Id); });
        $("#btnRightSidebarTodoImportant").click(() => { State.Current().TodoImportantChanged(State.Current().CurrentTodo.Id); });
    }

    public DrawTLists(): void {
        $('#NavigationBaseTLists').empty()
        $('#NavigationUserTLists').empty()
        this.CreateTList(State.Current().TodayTodos, "lstTodayTodos", "NavigationBaseTLists");
        this.CreateTList(State.Current().ImportantTodos, "lstImportantTodos", "NavigationBaseTLists");
        this.CreateTList(State.Current().PlannedTodos, "lstPlannedTodos", "NavigationBaseTLists");
        this.CreateTList(State.Current().UndividedTodos, "lstUndividedTodos", "NavigationBaseTLists");

        for (let item of State.Current().TLists)
            this.CreateTList(item, `lstItem${item.Id}`, "NavigationUserTLists");
    }

    public CreateTList(tList:TList, lstId:string, container:string): void {
        tList.GetUnfinishedTodosCount((result) => {
            let markup: JQuery = $(`
            <li id='${lstId}' class='LeftSidebar-item ${State.Current().CurrentTList == tList ? "is-selected" : ""}'>
                <span class='LeftSidebar-itemIcon ${tList.Icon}'></span>
                <span class='LeftSidebar-itemText'>${tList.Name}</span>
                <span class='LeftSidebar-itemCount'>${result}</span>
            </li>
            `);
            markup.appendTo(`#${container}`);
            $(`#${lstId}`).click(() => { State.Current().SwitchTList(tList); });
        });
    }

    public SetRightSidebar(): void {
        $("#chbRightSidebarTodoDone").prop("checked", State.Current().CurrentTodo.Done);
        $("#txtRightSidebarTodoText").val(State.Current().CurrentTodo.Text);
        this.SetRightSidebarTodoImportant(State.Current().CurrentTodo.Important);
        let endDate:Date = new Date(State.Current().CurrentTodo.EndDate);
        $("#dtpRightSidebarTodoEndDate").val(State.Current().CurrentTodo.EndDate != DateMaxValue.toDateString() ? `${this.formatDate(endDate)}` : "");
        $("#txaRightSidebarTodoNote").val(State.Current().CurrentTodo.Note);
        $("#lblRightSidebarTodoCreateDate").text(State.Current().CurrentTodo.CreateDate);
    }

    public formatDate(date): string {
        let d:Date = new Date(date)
        let month:string = '' + (d.getMonth() + 1)
        let day:string = '' + d.getDate()
        let year:number = d.getFullYear()
        if (month.length < 2)
            month = '0' + month
        if (day.length < 2)
            day = '0' + day
        return [year, month, day].join('-');
    }

    public DrawMain(): void {
        $('#TodoDividerContainer').empty();
        if (this.SearchView)
            this.CreateTodoDivider("dvcSearched", "Vyhledáno", true, State.Current().UnfinishedTodos, true);
        else {
            this.SetHeader()
            this.CreateTodoDivider("dvcUnfinished", "Nedokončeno", State.Current().CurrentTList.UnfinishedOpened, State.Current().UnfinishedTodos, true);
            this.CreateTodoDivider("dvcFinished", "Dokončeno", State.Current().CurrentTList.FinishedOpened, State.Current().FinishedTodos, false);
        }
    }

    public CreateTodoDivider(id: string, text: string, opened: boolean, todos: Todo[], unfinishedButton: boolean): void {
        let dividerContainer: string = `${id}TodoContainer`;
        let markup: JQuery = $(`
        <div id='${id}' class='TaskDivider'>
            <span class='TaskDivider-button ${opened ? "bi bi-chevron-down" : "bi bi-chevron-right"}'></span>
            <span>${text}</span>
            <span>${todos != undefined && todos.length > 0 ? todos.length : ""}</span>
        </div>
        <div id='${dividerContainer}'></div>
        `)
        markup.appendTo(`#TodoDividerContainer`);
        $(`#${id}`).click(() => { State.Current().TodoDividerClick(unfinishedButton); });
        if (todos != undefined && opened) {
            for (let todo of todos)
                this.CreateTodo(todo, dividerContainer);
        }
    }

    public CreateTodo(todo:Todo, container:string): void {
        let markup: JQuery = $(`
        <div class='TaskBlock ${State.Current().CurrentTodo == todo ? "is-selected" : ""}'>
            <div class='TaskBlock-check'>
                <div class='CheckboxContainer'>
                    <div class='Round'>
                        <input id='chbTaskDone${todo.Id}' ${todo.Done == true ? "checked" : ""} type='checkbox' />
                        <label for='chbTaskDone${todo.Id}'></label>
                    </div>
                </div>
            </div>
            <span id='lblTaskText${todo.Id}' class='TaskBlock-text'>${todo.Text}</span>
            <span id='lblTaskImportant${todo.Id}' class='TaskBlock-important ${todo.Important ? "is-checked" : ""}'>
                <i class='TaskBlock-importantIcon bi-star-fill'></i>
            </span>
        </div>
        `);
        markup.appendTo(`#${container}`);
        $(`#chbTaskDone${todo.Id}`).click(() => { State.Current().TodoDoneChanged(todo.Id); });
        $(`#lblTaskText${todo.Id}`).click(() => { State.Current().TodoSelected(todo.Id); });
        $(`#lblTaskImportant${todo.Id}`).click(() => { State.Current().TodoImportantChanged(todo.Id); });
    }

    public DrawTSteps(): void {
        $('#TStepContainer').empty();
        for (let step of State.Current().TSteps)
            this.CreateTStep(step);
    }

    public CreateTStep(step: TStep): void {
        let markup: JQuery = $(`
        <div class='RightSidebar-step'>
            <div class='RightSidebar-taskCheck'>
                <div class='CheckboxContainer'>
                    <div class='Round'>
                        <input id='chbStepDone${step.Id}' ${step.Done == true ? "checked" : ""} type='checkbox' />
                        <label for='chbStepDone${step.Id}'></label>
                    </div>
                </div>
            </div>
            <span class='RightSidebar-taskText TextBox'>
                <input id='txtStepText${step.Id}' class='TextBox-input' value='${step.Text}' type='text' />
            </span>
            <button id='btnStepDelete${step.Id}' class='Button RightSidebar-taskDelete' title='Smazat krok'>
                <i class='bi bi-trash'></i>
            </button>
        </div>
        `)
        markup.appendTo("#TStepContainer");
        $(`#chbStepDone${step.Id}`).click(() => { State.Current().TStepCheckedChanged(step.Id); });
        $(`#txtStepText${step.Id}`).change((e) => { State.Current().TStepTextChanged(e.target, step.Id); });
        $(`#btnStepDelete${step.Id}`).click(() => { State.Current().RemoveTStep(step.Id); });
    }

    public SetHeader(): void {
        $("#lblHeaderTListIcon").removeAttr('class');
        $("#lblHeaderTListIcon").addClass(State.Current().CurrentTList.Icon);
        if (this.SearchView)
            this.ShowSearchView();
        else
            this.HideSearchView();
        $("#txtHeaderTListName").prop("readonly", State.Current().CurrentTList.Id == UndividedTListId);
        $("#txtHeaderTListName").val(State.Current().CurrentTList.Name);
        if (State.Current().CurrentTList.Id != UndividedTListId)
            $("#TListControls").removeClass("u-d-none");
        else
            $("#TListControls").addClass("u-d-none");
        $("#txtSearchText").val("");
    }

    public SetColorMode(): void {
        if (Settings.Current().GetDarkMode() == "true") {
            if ($("#Page").hasClass("LightMode"))
                $("#Page").removeClass("LightMode");
            $("#Page").addClass("DarkMode");
        }
        else {
            if ($("#Page").hasClass("DarkMode"))
                $("#Page").removeClass("DarkMode");
            $("#Page").addClass("LightMode");
        }
    }

    public GetContentCoverStyle(): void {
        if (this.IsSmallScreen() && (this.LeftSidebarStatus == SidebarState.Opened || this.RightSidebarStatus == SidebarState.Opened))
            $("#MainCover").addClass("Main-coverContent");
        else
            $("#MainCover").removeClass("Main-coverContent");
    }

    public SetRightSidebarTodoImportant(todo): void {
        if (todo.Important)
            $("#btnRightSidebarTodoImportant").addClass("is-checked");
        else
            $("#btnRightSidebarTodoImportant").removeClass("is-checked");
    }

    public OutsideClick(): void {
        if (this.IsSmallScreen()) {
            $("#LeftSidebar").addClass("is-hidden");
            this.LeftSidebarStatus = SidebarState.Closed;
            this.GetContentCoverStyle();
        }
    }

    public ToggleLeftSidebar(): void {
        this.LeftSidebarStatus = this.ToggleSidebar(this.LeftSidebarStatus, "#LeftSidebar");
        this.GetContentCoverStyle();
    }

    public ToggleRightSidebar(): void {
        this.RightSidebarStatus = this.ToggleSidebar(this.RightSidebarStatus, "#RightSidebar");
        this.GetContentCoverStyle();
    }

    public ToggleSidebar(sidebarStatus: SidebarState, id:string): SidebarState {
        switch (sidebarStatus) {
            case SidebarState.Closed:
                $(id).removeClass("is-hidden");
                return SidebarState.Opened;
            default:
                $(id).addClass("is-hidden");
                return SidebarState.Closed;
        }
    }

    public IsSmallScreen(): boolean {
        return $("#Page").innerWidth() < 992;
    }

    public DownloadFile(fileBase64:string): void {
        fetch(`data:application/json;base64,${fileBase64}`)
            .then(response => response.blob())
            .then(blob => {
                let link = window.document.createElement("a");
                link.href = window.URL.createObjectURL(blob);
                link.download = "TodoExport.json";
                document.body.appendChild(link);
                link.click();
                document.body.removeChild(link);
            });
    }

    public ShowSearchView(): void {
        $("#lblHeaderTListIcon").addClass("u-d-none");
        $("#txtHeaderTListName").addClass("u-d-none");
        $("#TListControls").addClass("u-d-none");
        $("#SearchViewEditor").removeClass("u-d-none");
        $("#AddTodoContainer").addClass("u-d-none");
        $("#txtSearchText").val("");
        this.SearchView = true
    }

    public HideSearchView(): void {
        $("#lblHeaderTListIcon").removeClass("u-d-none");
        $("#txtHeaderTListName").removeClass("u-d-none");
        $("#TListControls").removeClass("u-d-none");
        $("#SearchViewEditor").addClass("u-d-none");
        $("#AddTodoContainer").removeClass("u-d-none");
        this.SearchView = false
    }

    public GetSearchText(): string {
        return $("#txtSearchText").val();
    }

    public ClearNewTListName(): void {
        $("#txtNewTListText").val("");
    }

    public GetNewTListName(): string {
        return $("#txtNewTListText").val();
    }

    public ClearNewTodoText(): void {
        $("#txtNewTodoText").val("");
    }

    public GetNewTodoText(): string {
        return $("#txtNewTodoText").val();
    }

    public ClearNewTStepText(): void {
        $("#txtNewTStepText").val("");
    }

    public GetNewTStepText(): string {
        return $("#txtNewTStepText").val();
    }

    public ShowTextDialog(title: string, text: string): void {
        this.DialogContent = DialogContent.Text;
        $("#lblDialogHead").text(title);
        $("#DialogTextContainer").text(text);
        this.ShowDialog();
    }

    public ShowSettingsDialog(): void {
        this.DialogContent = DialogContent.Settings;
        $("#lblDialogHead").text("Nastavení");
        this.ShowDialog();
    }

    public ShowTListSettingsDialog(): void {
        this.DialogContent = DialogContent.TListSettings;
        $("#lblDialogHead").text("Nastavení seznamu: " + State.Current().CurrentTList.Name);
        this.ShowDialog();
    }

    public ShowDialog(): void {
        switch (this.DialogContent) {
            case DialogContent.Settings:
                $("#DialogTextContainer").addClass("u-d-none");
                $("#DialogSettingsContainer").removeClass("u-d-none");
                $("#DialogTListSettingsContainer").addClass("u-d-none");
                $("#chbDarkMode").prop("checked", Settings.Current().GetDarkMode() == "true");
                break;
            case DialogContent.TListSettings:
                $("#DialogTextContainer").addClass("u-d-none");
                $("#DialogSettingsContainer").addClass("u-d-none");
                $("#DialogTListSettingsContainer").removeClass("u-d-none");
                $("#slbOrderTodo").val(State.Current().CurrentTList.OrderTodo);
                $("#slbIcon").val(State.Current().CurrentTList.Icon);
                break;
            default:
                $("#DialogTextContainer").removeClass("u-d-none");
                $("#DialogSettingsContainer").addClass("u-d-none");
                $("#DialogTListSettingsContainer").addClass("u-d-none");
                break;
        }
        $("#Dialog").addClass("is-showed");
    }

    public HideDialog(): void {
        $("#Dialog").removeClass("is-showed");
    }
}